vochain/account.go

```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
+	binary.LittleEndian.PutUint64(b, faucetPayload.Identifier)
+	key := from.Bytes()
+	key = append(key, b...)
+	keyHash := crypto.Sha256(key)
+	if err := v.SetFaucetNonce(keyHash); err != nil {
+		return err
 	}
-	log.Debugf("creating account %s with infoURI %s balance %d and delegates %+v",
-		accountAddress.String(),
-		acc.InfoURI,
-		acc.Balance,
-		printPrettierDelegates(acc.DelegateAddrs),
+	log.Debugf("consuming faucet payload created by %s with amount %d and identifier %d (keyHash: %x)",
+		from.String(),
+		faucetPayload.Amount,
+		faucetPayload.Identifier,
+		keyHash,
 	)
-	return v.SetAccount(accountAddress, acc)
+	return nil
 }
 
-// ConsumeFaucetPayload consumes a given faucet payload and sends the given amount of tokens to
-// the address pointed by the payload.
-func (v *State) ConsumeFaucetPayload(from common.Address, faucetPayload *models.FaucetPayload, txType models.TxType) error {
-	// check faucet payload
-	if faucetPayload == nil {
-		return fmt.Errorf("faucet payload is nil")
+// CreateAccountTxCheck is an abstraction of the ABCI CheckTx method for creating an account
+func CreateAccountTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) error {
+	if vtx == nil || txBytes == nil || signature == nil || state == nil {
+		return fmt.Errorf("invalid parameters provided, cannot check create account tx")
 	}
-	// check from account
-	if from == types.EthereumZeroAddress {
-		return fmt.Errorf("invalid from account")
+	tx := vtx.GetSetAccount()
+	if tx == nil {
+		return fmt.Errorf("invalid tx")
+	}
+	if tx.Txtype != models.TxType_CREATE_ACCOUNT {
+		return fmt.Errorf("invalid tx type, expected %s, got %s", models.TxType_CREATE_ACCOUNT, tx.Txtype)
 	}
-	var accFrom, accTo *Account
-	var err error
-	// get from account
-	accFrom, err = v.GetAccount(from, false)
+	pubKey, err := ethereum.PubKeyFromSignature(txBytes, signature)
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot extract public key from signature: %w", err)
 	}
-	if accFrom == nil {
-		return ErrAccountNotExist
+	txSenderAddress, err := ethereum.AddrFromPublicKey(pubKey)
+	if err != nil {
+		return fmt.Errorf("cannot extract address from public key: %w", err)
 	}
-	// get to account
-	accToAddr := common.BytesToAddress(faucetPayload.To)
-	accTo, err = v.GetAccount(accToAddr, false)
+	txSenderAcc, err := state.GetAccount(txSenderAddress, false)
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot get account: %w", err)
 	}
-	if accTo == nil {
-		return ErrAccountNotExist
+	if txSenderAcc != nil {
+		return ErrAccountAlreadyExists
 	}
-
-	// get burn account
-	burnAcc, err := v.GetAccount(BurnAddress, false)
+	infoURI := tx.GetInfoURI()
+	if len(infoURI) > types.MaxURLLength {
+		return ErrInvalidURILength
+	}
+	if len(tx.Delegates) != 0 {
+		delegatesMap := make(map[common.Address]bool)
+		for _, v := range tx.Delegates {
+			delegateAddress := common.BytesToAddress(v)
+			if delegateAddress == types.EthereumZeroAddress {
```

Comment:
I'd not consider EthereumZeroAddress as an special address. Its just too difficult to find it.